Apps/NotSensorLogger2/app/src/main/java/com/example/notsensorlogger2/MainActivity.java

```diff
@@ -0,0 +1,353 @@
+
+        sensorManager = (SensorManager) getSystemService(SENSOR_SERVICE);
+        accelerometerHandler = new AccelerometerHandler(sensorManager);
+        gyroscopeHandler = new GyroscopeHandler(sensorManager);
+        magnetometerHandler = new MagnetometerHandler(sensorManager);
+        lightSensorHandler = new LightSensorHandler(sensorManager);
+        humiditySensorHandler = new RelativeHumiditySensorHandler(sensorManager);
+        pressureSensorHandler = new PressureSensorHandler(sensorManager);
+        gravitySensorHandler = new GravitySensorHandler(sensorManager);
+        locationTracker = new LocationHandler((Context) this);
+        soundLevelHandler = new SoundLevelHandler(this, sensorManager);
+
+        // After initializing all sensors and handlers, call the setup method for RecyclerView
+        setupSensorList();
+
+        Button stopDataButton = findViewById(R.id.start_recording_button);
+        stopDataButton.setText(R.string.start_button_text);
+
+        isDataCollectionActive = false;
+
+        stopDataButton.setOnClickListener(v -> {
+            if (!isDataCollectionActive) {
+                startDataCollection();
+                stopDataButton.setText(R.string.stop_button_text);
+                isDataCollectionActive = true;
+            } else {
+                stopDataCollection();
+                stopDataButton.setText(R.string.start_button_text);
+                isDataCollectionActive = false;
+            }
+        });
+
+        Button btnOpenActivity2 = findViewById(R.id.open_activity2_button);
+        btnOpenActivity2.setOnClickListener(v -> {
+            Intent intent = new Intent(this, MainActivity2.class);
+            startActivity(intent);
+        });
+
+        sendDataRunnable = new Runnable() {
+            @Override
+            public void run() {
+                sendSensorData();
+                handler.postDelayed(this, SEND_INTERVAL);
+            }
+        };
+    }
+
+    /**
+     * Resumes the activity, checks necessary permissions, and resumes data collection if it was active before pausing.
+     */
+    @Override
+    protected void onResume() {
+        super.onResume();
+        checkAndRequestAudioPermission();
+        if (isDataCollectionActive) {
+            startDataCollection();
+        }
+    }
+
+    /**
+     * Pauses the activity and stops data collection if it was active.
+     */
+    @Override
+    protected void onPause() {
+        super.onPause();
+        if (isDataCollectionActive) {
+            stopDataCollection();
+        }
+    }
+
+    /**
+     * Starts the data collection from all active sensors and schedules periodic data sending to the server.
+     */
+    private void startDataCollection() {
+        messageId = 1;
+        for (SensorItem item : sensorItems) {
+            if (item.isToggled() && item.getHandler() != null) {
+                item.getHandler().start();
+            }
+        }
+        if (!isDataCollectionActive) {
```

Comment:
Is this check necessary? 
Found a bug in the control flow with this variable in line 202:
In onResume, `startDataCollection` is only called when `isDataCollectionActive` is true. But in `startDataCollection`, `sendDataRunnable` is only posted when `isDataCollectionActive` is false.

Please reanalyse the flow, and remove the variable if it is not necessary here.

Reply:
It helps as sendDataRunnable gets triggered unexpectedly and unwillingly without this check.

Reply:
In what case will it be triggered unexpectedly? If it is triggered unexpectedly and unwillingly, should it be considered as a bug and be fixed?

Reply:
Changed it up so that the check is removed, and works normally